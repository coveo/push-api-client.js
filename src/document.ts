import {PermissionLevel, PermissionSet} from '@coveo/platform-client';

/**
 * A simple permission model. See https://docs.coveo.com/en/107 for more information.
 */
// export interface PermissionIdentityModel {
//   /**
//    * The name of the security identity.
//    *
//    * Examples:
//    *  - `asmith@example.com`
//    *  - `SampleTeam2`
//    */
//   identity: string;
//   /**
//    * The type of the identity.
//    * Valid values:
//    * - `UNKNOWN`
//    * - `USER` : Defines a single user.
//    * - `GROUP` : Defines an existing group of identities within the indexed system. Individual members of this group can be of any valid identity Type (USER, GROUP, or VIRTUAL_GROUP).
//    * - `VIRTUAL_GROUP` : Defines a group that doesn't exist within the indexed system. Mechanically, a `VIRTUAL_GROUP` is identical to a `GROUP`.
//    */
//   identityType: SinglePermissionIdentityType;
//   /**
//    * The security identity provider through which the security identity is updated.
//    *
//    * Defaults to the first security identity provider associated with the target Push source.
//    */
//   securityProvider?: string;
// }

type BaseMetadataValue = string | string[] | number | number[] | boolean;
type DictionaryMetadataValue = Record<string, BaseMetadataValue>;
export type MetadataValue = BaseMetadataValue | DictionaryMetadataValue;
export type Metadata = Record<string, MetadataValue>;

/**
 * The compression type that was applied to your compressed document.
 */
export type CompressionType =
  | 'UNCOMPRESSED'
  | 'DEFLATE'
  | 'GZIP'
  | 'LZMA'
  | 'ZLIB';

/**
 * A Coveo document.
 */
export interface Document {
  /**
   * The Uniform Resource Identifier (URI) that uniquely identifies the document in a Coveo index.
   *
   * Examples:
   * - `http://www.example.com/`
   * - `file://folder/text.txt`
   */
  uri: string;
  /**
   * The title of the document.
   */
  title: string;
  /**
   * The clickable URI associated with the document.
   */
  clickableUri?: string;
  /**
   * The author of the document.
   */
  author?: string;
  /**
   * The date of the document, represented as an ISO string.
   *
   * Optional, will default to indexation date.
   */
  date?: string;
  /**
   * The modified date of the document, represented as an ISO string.
   *
   * Optional, will default to indexation date.
   */
  modifiedDate?: string;
  /**
   * The permanent identifier of a document that does not change over time.
   *
   * Optional, will be derived from the document URI.
   */
  permanentId?: string;
  /**
   * The unique identifier (URI) of the parent item.
   *
   * Specifying a value for this key creates a relationship between the attachment item (child) and its parent item.
   *
   * This value also ensures that a parent and all of its attachments will be routed in the same index slice.
   */
  parentId?: string;
  /**
   * The textual (non-binary) content of the item.
   *
   * Whenever you're pushing a compressed binary item (such as XML/HTML, PDF, Word, or binary), you should use the CompressedBinaryData or CompressedBinaryDataFileId attribute instead, depending on the content size.
   *
   * Accepts 5 MB or less of uncompressed textual data.
   *
   * See https://docs.coveo.com/en/73 for more information.
   *
   * Example: `This is a simple string that will be used for searchability as well as to generate excerpt and summaries for the document.`
   */
  data?: string;
  /**
   * The original binary item content, compressed using one of the supported compression types (Deflate, GZip, LZMA, Uncompressed, or ZLib), and then Base64 encoded.
   *
   * You can use this parameter when you're pushing a compressed binary item (such as XML/HTML, PDF, Word, or binary) whose size is less than 5 MB.
   *
   * Whenever you're pushing an item whose size is 5 MB or more, use the CompressedBinaryDataFileIdproperty instead.
   *
   * If you're pushing less than 5 MB of textual (non-binary) content, you can use the data property instead.
   *
   * See https://docs.coveo.com/en/73 for more information.
   */
  compressedBinaryData?: {
    /**
     * The compression type that was applied to your document.
     */
    compressionType: CompressionType;
    /**
     * The base64 encoded binary data.
     *
     * Example: `eJxzrUjMLchJBQAK4ALN`
     */
    data: string;
  };
  /**
   * The metadata key-value pairs for a given document.
   *
   * Each metadata in the document must be unique.
   *
   * Metadata are case-insensitive (e.g., the Push API considers mykey, MyKey, myKey, MYKEY, etc. as identical).
   *
   * See https://docs.coveo.com/en/115 for more information.
   */
  metadata?: Metadata;
  /**
   * The list of permission sets for this item.
   *
   * This is useful when item based security is required (i.e., when security isn't configured at the source level).
   *
   * See https://docs.coveo.com/en/107 for more information.
   */
  permissions?: Array<PermissionSet | PermissionLevel>;
  /**
   * The file extension of the data you're pushing.
   *
   * This is useful when pushing a compressed item. The converter uses this information to identify how to correctly process the item.
   *
   * Values must include the preceding . character.
   *
   * Example: `.html`
   */
  fileExtension?: string;
}
