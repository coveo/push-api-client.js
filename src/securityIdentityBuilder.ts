// eslint-disable-next-line @typescript-eslint/no-unused-vars
import type {PermissionSetBuilder} from './permissionSetBuilder';
import {Permission, PermissionIdentityType} from '@coveo/platform-client';

/**
 * Build a security identity. See {@link Permission}.
 */
export interface SecurityIdentityBuilder {
  build(): Permission | Permission[];
}

/**
 * Build any {@link PermissionIdentityType}.
 *
 * Instead of using this class directly, use one of:
 * - {@link UserSecurityIdentityBuilder}
 * - {@link GroupSecurityIdentityBuilder}
 * - {@link VirtualGroupSecurityIdentityBuilder}
 */
export class AnySecurityIdentityBuilder implements SecurityIdentityBuilder {
  private securityIdentity: Permission;
  /**
   * @param identityType  The type of the identity.
   * Valid values:
   * - `UNKNOWN`
   * - `USER` : Defines a single user.
   * - `GROUP` : Defines an existing group of identities within the indexed system. Individual members of this group can be of any valid identity Type (USER, GROUP, or VIRTUAL_GROUP).
   * - `VIRTUAL_GROUP` : Defines a group that doesn't exist within the indexed system. Mechanically, a `VIRTUAL_GROUP` is identical to a `GROUP`.
   *
   * @param identity The name of the security identity.
   * Examples:
   * - `asmith@example.com`
   * - `SampleTeam2`
   *
   * @param securityProvider The security identity provider through which the security identity is updated. Defaults to the first security identity provider associated with the target source.
   */
  constructor(
    identityType: PermissionIdentityType,
    identity: string,
    securityProvider?: string
  ) {
    this.securityIdentity = {
      identityType,
      identity,
      ...(securityProvider && {securityProvider}),
    };
  }

  /**
   * Build and return the security identity.
   * @returns
   */
  public build() {
    return this.securityIdentity;
  }
}

/**
 * Build a security identity of type `USER`.
 *
 * Typically used in conjunction with {@link PermissionSetBuilder.withAllowedPermissions} or {@link PermissionSetBuilder.withDeniedPermissions}.
 *
 * See {@link Permission}.
 */
export class UserSecurityIdentityBuilder implements SecurityIdentityBuilder {
  /**
   * Pass either a single user, or an array of user to create multiple identities.
   * @param {(string | string[])} user  The user identity or identities
   * @param {string} [securityProvider='Email Security Provider'] The security identity provider through which the security identity is updated. Defaults to the first security identity provider associated with the target source.
   */
  constructor(
    private user: string | string[],
    private securityProvider: string = 'Email Security Provider'
  ) {}

  /**
   * Build and return the security identity or identities.
   * @returns
   */
  public build() {
    if (Array.isArray(this.user)) {
      return this.user.map((u) =>
        new AnySecurityIdentityBuilder(
          PermissionIdentityType.User,
          u,
          this.securityProvider
        ).build()
      );
    }
    return new AnySecurityIdentityBuilder(
      PermissionIdentityType.User,
      this.user,
      this.securityProvider
    ).build();
  }
}

/**
 * Build a security identity of type `GROUP`.
 *
 * Typically used in conjunction with {@link PermissionSetBuilder.withAllowedPermissions} or {@link PermissionSetBuilder.withDeniedPermissions}.
 *
 * See {@link Permission}.
 */
export class GroupSecurityIdentityBuilder implements SecurityIdentityBuilder {
  /**
   * Pass either a single `group`, or an array of `group` to create multiple identities.
   * @param {(string | string[])} group
   * @param {string} [securityProvider] The security identity provider through which the security identity is updated. Defaults to the first security identity provider associated with the target source.
   */
  constructor(
    private group: string | string[],
    private securityProvider?: string
  ) {}

  /**
   * Build and return the security identity or identities.
   * @returns
   */
  public build() {
    if (Array.isArray(this.group)) {
      return this.group.map((g) =>
        new AnySecurityIdentityBuilder(
          PermissionIdentityType.Group,
          g,
          this.securityProvider
        ).build()
      );
    }
    return new AnySecurityIdentityBuilder(
      PermissionIdentityType.Group,
      this.group,
      this.securityProvider
    ).build();
  }
}
/**
 * Build a security identity of type `VIRTUAL_GROUP`.
 *
 * Typically used in conjunction with {@link PermissionSetBuilder.withAllowedPermissions} or {@link PermissionSetBuilder.withDeniedPermissions}.
 *
 * See {@link Permission}.
 */
export class VirtualGroupSecurityIdentityBuilder
  implements SecurityIdentityBuilder
{
  /**
   * Pass either a single `virtualGroup`, or an array of `virtualGroup` to create multiple identities.
   * @param {(string | string[])} virtualGroup
   * @param {string} [securityProvider] The security identity provider through which the security identity is updated. Defaults to the first security identity provider associated with the target source.
   */
  constructor(
    private virtualGroup: string | string[],
    private securityProvider?: string
  ) {}

  /**
   * Build and return the security identity or identities.
   * @returns
   */
  public build() {
    if (Array.isArray(this.virtualGroup)) {
      return this.virtualGroup.map((vg) =>
        new AnySecurityIdentityBuilder(
          PermissionIdentityType.VirtualGroup,
          vg,
          this.securityProvider
        ).build()
      );
    }
    return new AnySecurityIdentityBuilder(
      PermissionIdentityType.VirtualGroup,
      this.virtualGroup,
      this.securityProvider
    ).build();
  }
}
