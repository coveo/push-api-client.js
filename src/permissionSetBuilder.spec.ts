import {PermissionSetModel} from './document';
import {SecurityIdentityBuilder} from './securityIdentityBuilder';

export class PermissionSetBuilder {
  private permissionSet: Partial<PermissionSetModel> = {};

  public constructor() {}

  /**
   * Set allowed identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withAllowedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'allowedPermissions');
    return this;
  }

  /**
   * Set denied identities on the document. See {@link Document.permissions}
   * @param securityIdentityBuilder
   * @returns
   */
  public withDeniedPermissions(
    securityIdentityBuilder: SecurityIdentityBuilder
  ) {
    this.setPermission(securityIdentityBuilder, 'deniedPermissions');
    return this;
  }

  /**
   * Set allowAnonymous for permissions on the document. See {@link Document.permissions}
   * @param allowAnonymous
   * @returns
   */
  public withAllowAnonymousUsers(allowAnonymous: boolean) {
    this.permissionSet.allowAnonymous = allowAnonymous;
    return this;
  }

  private setPermission(
    securityIdentityBuilder: SecurityIdentityBuilder,
    permissionSection: 'allowedPermissions' | 'deniedPermissions'
  ) {
    const identities = securityIdentityBuilder.build();
    if (!this.permissionSet[permissionSection]) {
      this.permissionSet[permissionSection] = [];
    }
    if (Array.isArray(identities)) {
      this.permissionSet[permissionSection] =
        this.permissionSet[permissionSection]?.concat(identities);
    } else {
      this.permissionSet[permissionSection]?.push(identities);
    }
  }
}
