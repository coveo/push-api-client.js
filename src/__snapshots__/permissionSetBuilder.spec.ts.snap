// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PermissionSetBuilder should marshal allowedPermissions in multiple #withAllowedPermissions 1`] = `
{
  "allowAnonymous": false,
  "allowedPermissions": [
    {
      "identity": "max@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "sue@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "bob@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "SampleGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "SpecialGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "my_group",
      "identityType": "GROUP",
    },
  ],
  "deniedPermissions": [],
}
`;

exports[`PermissionSetBuilder should marshal both allowed and denied permissions  1`] = `
{
  "allowAnonymous": false,
  "allowedPermissions": [
    {
      "identity": "SampleGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "SpecialGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "my_group",
      "identityType": "GROUP",
    },
  ],
  "deniedPermissions": [
    {
      "identity": "max@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "sue@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "bob@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
  ],
}
`;

exports[`PermissionSetBuilder should marshal deniedPermissions in multiple #withDeniedPermissions 1`] = `
{
  "allowAnonymous": false,
  "allowedPermissions": [],
  "deniedPermissions": [
    {
      "identity": "max@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "sue@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "bob@foo.com",
      "identityType": "USER",
      "securityProvider": "Email Security Provider",
    },
    {
      "identity": "SampleGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "SpecialGroup",
      "identityType": "VIRTUAL_GROUP",
    },
    {
      "identity": "my_group",
      "identityType": "GROUP",
    },
  ],
}
`;
